src/Microsoft.Health.Fhir.CosmosDb/Features/Storage/CosmosFhirDataStore.cs

```diff
@@ -126,7 +126,7 @@ public sealed class CosmosFhirDataStore : IFhirDataStore, IProvideCapability
                 switch (dce.GetSubStatusCode())
                 {
                     case HttpStatusCode.PreconditionFailed:
-                        throw new ResourceConflictException(weakETag);
+                        throw new PreconditionFailedException(string.Format(Core.Resources.ResourceVersionConflict, weakETag?.VersionId));
                     case HttpStatusCode.NotFound:
                         if (cosmosWrapper.IsDeleted)
                         {
```

Comment:
@brendankowitz the spec doesn't seem to clearly outline how to handle the scenario where an incorrect if-match header is provided for a resource that does not exist. My interpretation would be that we should return a 404 error code on line 138 below, but I could also see how 409/412 might make sense.

Reply:
I think that interpretation is correct. If the resource doesn't exist, we shoudl fail with 404. Only if the resource exists, then we will take if-match header into consideration.

Reply:
Mini update, I came across this comment in the spec:

> The update interaction creates a new current version for an existing resource or creates an initial version if no resource already exists for the given id.

So just to clarify, if a correct if-match header is provided in an update call for a resource that does not exist, a new resource should be created. The if-match header would have to specify a version of 1 in this case.

If an incorrect if-match header is provided in an update call for resource that does not exist, I'm not sure if I should return 412 (for R4) and 409 (for Stu3) or if I should ignore the if-match header and create a resource with version 1.

Let me know what you think @jackliums @brendankowitz